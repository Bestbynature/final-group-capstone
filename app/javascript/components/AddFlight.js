import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import {
  postFlight, setPicture, setAvailableSlots, setBasePrice, setName,
} from '../redux/flights/flightsSlice';

const AddFlight = () => {
  const dispatch = useDispatch();
  const history = useNavigate();

  const {
    name, picture, basePrice, availableSlots,
    userId,
  } = useSelector((store) => store.flights);

  const handleSubmit = (e) => {
    e.preventDefault();

    const formData = new FormData();
    formData.append('flight[name]', name);
    formData.append('flight[picture]', picture);
    formData.append('flight[base_price]', basePrice);
    formData.append('flight[available_slots]', availableSlots);
    formData.append('flight[reserved]', 'false');
    formData.append('flight[user_id]', userId);

    dispatch(postFlight(formData));
    history('/');
  };

  return (

    <div className="modal fade" id="addModal" tabIndex="-1" aria-labelledby="addModalLabel" aria-hidden="true">
      <form onSubmit={handleSubmit}>
        <div className="modal-dialog">
          <div className="modal-content">
            <div className="modal-header">
              <h1 className="modal-title" id="addModalLabel">Add a New Flight</h1>
              <button type="button" className="btn-close" data-bs-dismiss="modal" aria-label="Close" />
            </div>
            <div className="modal-body">
              <input
                className="form-control mt-3"
                type="text"
                value={name}
                placeholder="Flight Name"
                onChange={(e) => dispatch(setName(e.target.value))}
                required
                data-testid="flight-name-input"

              />

              <input
                className="form-control mt-3"
                type="text"
                value={picture}
                placeholder="Picture URL for the flight"
                onChange={(e) => dispatch(setPicture(e.target.value))}
                required
                data-testid="flight-picture-input"
              />

              <input
                className="form-control mt-3"
                type="number"
                value={basePrice}
                placeholder="Base Price for the destination"
                onChange={(e) => dispatch(setBasePrice(e.target.value))}
                required
                data-testid="flight-base-price-input"
              />

              <input
                className="form-control mt-3"
                type="number"
                value={availableSlots}
                placeholder="Available Slots for the destination"
                onChange={(e) => dispatch(setAvailableSlots(e.target.value))}
                required
                data-testid="flight-available-slots-input"
              />
            </div>
            <div className="modal-footer">
              <button type="button" className="btn btn-danger form-control mt-2" data-bs-dismiss="modal">Cancel</button>
              <button type="submit" className="btn add-submit form-control mt-3" data-bs-dismiss="modal" data-testid="add-flight-form">Create a Flight</button>
            </div>
          </div>
        </div>
      </form>
    </div>
  );
};

export default AddFlight;
