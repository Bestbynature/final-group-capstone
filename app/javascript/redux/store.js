import { configureStore } from '@reduxjs/toolkit';
import flightsReducer from './flights/flightsSlice';

const store = configureStore({
  reducer: {
    flights: flightsReducer,
  },
});

export default store;
