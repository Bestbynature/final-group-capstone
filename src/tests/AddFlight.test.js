import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AddFlight from '../../app/javascript/components/AddFlight';
import store from '../../app/javascript/redux/store';

describe('AddFlight', () => {
  it('renders add flight form correctly', () => {
    const details = renderer
      .create(
        <BrowserRouter>
          <Provider store={store}>
            <AddFlight />
          </Provider>
        </BrowserRouter>,
      )
      .toJSON();
    expect(details).toMatchSnapshot();
  });

  it('submits new flight form and dispatches action correctly', async () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <Provider store={store}>
          <AddFlight />
        </Provider>
      </BrowserRouter>,
    );

    // Fill in the form fields
    fireEvent.change(getByTestId('flight-name-input'), { target: { value: 'New Flight' } });
    fireEvent.change(getByTestId('flight-picture-input'), { target: { value: 'https://example.com/picture.jpg' } });
    fireEvent.change(getByTestId('flight-base-price-input'), { target: { value: '100' } });
    fireEvent.change(getByTestId('flight-available-slots-input'), { target: { value: '10' } });

    fireEvent.submit(getByTestId('add-flight-form'));
  });
});
