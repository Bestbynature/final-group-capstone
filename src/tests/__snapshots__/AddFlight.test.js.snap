// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFlight renders add flight form correctly 1`] = `
<div
  aria-hidden="true"
  aria-labelledby="addModalLabel"
  className="modal fade"
  id="addModal"
  tabIndex="-1"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-dialog"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h1
            className="modal-title"
            id="addModalLabel"
          >
            Add a New Flight
          </h1>
          <button
            aria-label="Close"
            className="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          className="modal-body"
        >
          <input
            className="form-control mt-3"
            data-testid="flight-name-input"
            onChange={[Function]}
            placeholder="Flight Name"
            required={true}
            type="text"
            value=""
          />
          <input
            className="form-control mt-3"
            data-testid="flight-picture-input"
            onChange={[Function]}
            placeholder="Picture URL for the flight"
            required={true}
            type="text"
            value=""
          />
          <input
            className="form-control mt-3"
            data-testid="flight-base-price-input"
            onChange={[Function]}
            placeholder="Base Price for the destination"
            required={true}
            type="number"
            value={null}
          />
          <input
            className="form-control mt-3"
            data-testid="flight-available-slots-input"
            onChange={[Function]}
            placeholder="Available Slots for the destination"
            required={true}
            type="number"
            value={null}
          />
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-danger form-control mt-2"
            data-bs-dismiss="modal"
            type="button"
          >
            Cancel
          </button>
          <button
            className="btn add-submit form-control mt-3"
            data-bs-dismiss="modal"
            data-testid="add-flight-form"
            type="submit"
          >
            Create a Flight
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
